File: agent-firewall-js/lib/policy.js
/**
 * Policy loader — reads a YAML policy file and returns a structured Policy object.
 * Port of proxy/app/policy.py to JavaScript.
 */

import { readFileSync } from "node:fs";
import { resolve } from "node:path";
import yaml from "js-yaml";

/**
 * @typedef {Object} SecretPattern
 * @property {string} name
 * @property {string} regex
 */

/**
 * @typedef {Object} ArgRule
 * @property {string} name
 * @property {string} reason
 * @property {string} regex
 * @property {string[]} tools
 */

/**
 * @typedef {Object} PolicyData
 * @property {number}          version
 * @property {string}          description
 * @property {string[]}        tool_allowlist
 * @property {string[]}        tool_denylist
 * @property {string[]}        egress_allowlist
 * @property {SecretPattern[]} secret_patterns
 * @property {ArgRule[]}       arg_rules
 * @property {string[]}        injection_phrases
 * @property {number}          injection_threshold
 * @property {number}          token_budget_per_session
 * @property {boolean}         fail_closed
 * @property {string}          block_message
 */

const DEFAULTS = {
  version: 1,
  description: "",
  tool_allowlist: [],
  tool_denylist: [],
  egress_allowlist: [],
  secret_patterns: [],
  arg_rules: [],
  injection_phrases: [],
  injection_threshold: 0.8,
  token_budget_per_session: 20000,
  fail_closed: true,
  block_message: "[Agent Firewall] Action blocked by policy.",
};

/**
 * Compile a regex string that may contain Python-style inline flags like (?i).
 * JavaScript doesn't support inline flag groups, so we extract them and pass
 * as RegExp constructor flags.
 * @param {string} pattern
 * @returns {RegExp}
 */
function compileRegex(pattern) {
  let flags = "";
  let cleaned = pattern;
  // Extract leading (?flags) group — Python uses (?i), (?s), (?m), etc.
  const inlineMatch = cleaned.match(/^\(\?([imsu]+)\)/);
  if (inlineMatch) {
    flags = inlineMatch[1];
    cleaned = cleaned.slice(inlineMatch[0].length);
  }
  return new RegExp(cleaned, flags);
}

export class Policy {
  constructor(raw = {}) {
    const d = { ...DEFAULTS, ...raw };
    this.version = d.version;
    this.description = d.description;
    this.tool_allowlist = d.tool_allowlist || [];
    this.tool_denylist = d.tool_denylist || [];
    this.egress_allowlist = d.egress_allowlist || [];
    this.secret_patterns = (d.secret_patterns || []).map((p) => ({
      name: p.name,
      regex: compileRegex(p.regex),
    }));
    this.arg_rules = (d.arg_rules || []).map((r) => ({
      name: r.name,
      reason: r.reason,
      regex: compileRegex(r.regex),
      tools: r.tools || ["*"],
    }));
    this.injection_phrases = d.injection_phrases || [];
    this.injection_threshold = d.injection_threshold;
    this.token_budget_per_session = d.token_budget_per_session;
    this.fail_closed = d.fail_closed;
    this.block_message = d.block_message;
  }

  /** @param {string} name */
  toolAllowed(name) {
    if (this.tool_denylist.includes(name)) return false;
    if (this.tool_allowlist.length === 0) return true;
    return this.tool_allowlist.includes(name);
  }

  /** @param {string} host */
  hostAllowed(host) {
    const h = host.toLowerCase().trim();
    for (const a of this.egress_allowlist) {
      const allowed = a.toLowerCase().trim();
      if (h === allowed || h.endsWith("." + allowed)) return true;
    }
    return false;
  }
}

/**
 * Load a policy from a YAML file path.
 * @param {string} filePath
 * @returns {Policy}
 */
export function loadPolicy(filePath) {
  const abs = resolve(filePath);
  const text = readFileSync(abs, "utf8");
  const raw = yaml.load(text) || {};
  return new Policy(raw);
}

File: voidhack_agent_firewall/policy.py
from __future__ import annotations

import re
from dataclasses import dataclass, field
from pathlib import Path

import yaml


@dataclass
class SecretPattern:
    name: str
    regex: str

    def compiled(self) -> re.Pattern[str]:
        return re.compile(self.regex)


@dataclass
class ArgRule:
    name: str
    reason: str
    regex: str
    tools: list[str] = field(default_factory=lambda: ["*"])

    def compiled(self) -> re.Pattern[str]:
        return re.compile(self.regex)

    def applies_to(self, tool: str) -> bool:
        return "*" in self.tools or tool in self.tools


@dataclass
class Policy:
    version: int = 1
    description: str = ""
    tool_allowlist: list[str] = field(default_factory=list)
    tool_denylist: list[str] = field(default_factory=list)
    egress_allowlist: list[str] = field(default_factory=list)
    secret_patterns: list[SecretPattern] = field(default_factory=list)
    arg_rules: list[ArgRule] = field(default_factory=list)
    injection_phrases: list[str] = field(default_factory=list)
    injection_threshold: float = 0.80
    token_budget_per_session: int = 20000
    fail_closed: bool = True
    block_message: str = "[Agent Firewall] Action blocked by policy."

    def tool_allowed(self, name: str) -> bool:
        if name in self.tool_denylist:
            return False
        if not self.tool_allowlist:
            return True
        return name in self.tool_allowlist

    def host_allowed(self, host: str) -> bool:
        host = host.lower().strip()
        for allowed in self.egress_allowlist:
            allowed_host = allowed.lower().strip()
            if host == allowed_host or host.endswith("." + allowed_host):
                return True
        return False

    def compiled_secrets(self) -> list[tuple[str, re.Pattern[str]]]:
        return [(pattern.name, pattern.compiled()) for pattern in self.secret_patterns]

    def compiled_arg_rules(self) -> list[tuple[ArgRule, re.Pattern[str]]]:
        return [(rule, rule.compiled()) for rule in self.arg_rules]


def load_policy(path: str | Path) -> Policy:
    raw = yaml.safe_load(Path(path).read_text(encoding="utf-8")) or {}
    return Policy(
        version=int(raw.get("version", 1)),
        description=str(raw.get("description", "")),
        tool_allowlist=list(raw.get("tool_allowlist") or []),
        tool_denylist=list(raw.get("tool_denylist") or []),
        egress_allowlist=list(raw.get("egress_allowlist") or []),
        secret_patterns=[
            SecretPattern(name=str(item["name"]), regex=str(item["regex"]))
            for item in raw.get("secret_patterns") or []
        ],
        arg_rules=[
            ArgRule(
                name=str(item["name"]),
                reason=str(item["reason"]),
                regex=str(item["regex"]),
                tools=list(item.get("tools") or ["*"]),
            )
            for item in raw.get("arg_rules") or []
        ],
        injection_phrases=list(raw.get("injection_phrases") or []),
        injection_threshold=float(raw.get("injection_threshold", 0.80)),
        token_budget_per_session=int(raw.get("token_budget_per_session", 20000)),
        fail_closed=bool(raw.get("fail_closed", True)),
        block_message=str(
            raw.get("block_message", "[Agent Firewall] Action blocked by policy.")
        ),
    )
